src/framework/preferences/js/SelfVoicingEnactor.js

```diff
@@ -92,45 +108,280 @@ var fluid_3_0_0 = fluid_3_0_0 || {};
+    fluid.prefs.enactor.selfVoicing.parsedToString = function (parsed) {
+        var words = fluid.transform(parsed, function (block) {
+            return block.word;
         });
+
+        return words.join("");
+    };
+
+    /**
+     * Parses the DOM element into data points to use for highlighting the text, and queues the text into the self
+     * voicing engine. The parsed data points are added as an array to the component's `parseQueue`
+     *
+     * @param {Component} that - the component
+     * @param {node} elm - The DOM node to read
+     */
+    fluid.prefs.enactor.selfVoicing.readFromDOM = function (that, elm) {
+        elm = $(elm);
+
+        // only execute if there are nodes to read from
+        if (elm.length) {
+            var parsedFromElm = fluid.prefs.enactor.selfVoicing.parse(elm[0]);
+            that.parseQueue.push(parsedFromElm);
+            that.events.onReadFromDOM.fire(parsedFromElm);
+            that.tts.queueSpeech(fluid.prefs.enactor.selfVoicing.parsedToString(parsedFromElm));
+        }
+    };
+
+    /**
+     * Returns the index of the closest data point from the parseQueue based on the boundary provided.
+     *
+     * @param {Array} parseQueue - An array data points generated from parsing a DOM structure
+     * @param {Number} currentIndex - The index into the paraseQueue to start searching from. The currentIndex will be
+     *                                constrained to the bounds of the parseQueue.
+     * @param {Number} boundary - The boundary value used to compare against the blockIndex of the parsed data points.
+     *
+     * @return {Number|undefined} - Will return the index of the closest data point in the parseQueue. If the boundary
+     *                               cannot be located within the parseQueue, `undefined` is returned.
+     */
+    fluid.prefs.enactor.selfVoicing.getClosestIndex = function (parseQueue, currentIndex, boundary) {
+        var maxIndex  = Math.max(parseQueue.length - 1, 0);
+        currentIndex = Math.max(Math.min(currentIndex, maxIndex), 0);
+
+        var nextIndex = currentIndex + 1;
+        var prevIndex = currentIndex - 1;
+
+        var currentBlockIndex = parseQueue[currentIndex].blockIndex;
+        var maxBoundary = parseQueue[maxIndex].blockIndex + parseQueue[maxIndex].word.length;
+
+
+        if (!fluid.isValue(boundary) || boundary < 0 || boundary > maxBoundary ) {
+            return undefined;
+        }
+
+        if (currentBlockIndex > boundary) {
+            return fluid.prefs.enactor.selfVoicing.getClosestIndex(parseQueue, prevIndex, boundary);
+        }
+
+        var isWithinNextBound = parseQueue[nextIndex] ? boundary < parseQueue[nextIndex].blockIndex : boundary <= maxBoundary;
+
+        if (currentBlockIndex === boundary || (currentIndex <= maxIndex && isWithinNextBound)) {
+            return currentIndex;
+        }
+
+        return fluid.prefs.enactor.selfVoicing.getClosestIndex(parseQueue, nextIndex, boundary);
```

Comment:
JavaScript still does not have "tail recursion elimination" so I suggest a clearer implementation would just be iterative - especially since there is a lot of setup that you do at the beginning of every function call which is simply duplicated each time

Reply:
@amb26 i'll explore that. However, i do have to down into sibling and descendant DOM Elements to pull out all of the data.

Reply:
@amb26 I posted a question in the fluid-work IRC channel about this. ( https://botbot.me/freenode/fluid-work/2018-04-12/?msg=98918443&page=1 ). 

In sure I haven't found a simple way to make this purely iterative, because we don't know ahead of time the nesting of DOM elements. I tried using querySelectorAll or jQuery's find method to get all of the DOM elements, but this would still require dealing with nested elements to preserve the order of the text.

Reply:
The DOM may be nested, but as far as I can see this method can be purely linear since it works only on the parseQueue structure which is an array, and the only varying coordinate is the 2nd "index" argument which is just an integer. As far as I can see this body could be replaced by one or two "while" loops, with a great increase in clarity

Reply:
@amb26 the method navigates through a nested structure to output a flattened one. I'm still struggling to think of how best to achieve this with the while loops given that the childnodes of an element may be text nodes or elements themselves. And these sub elements may have one or more sub elements and so on.